
<h1 align="center">Peer-graded Assignment: Segmenting and Clustering </h1>
<h1 align="center">Neighborhoods in Toronto</h1>

### Notebook contains the final assignment for week 3 of the Applied Data Science Capstone.

#### Greg Waitt
#### 3/24/21

### Required packages for assignment

In [1]:
import pandas as pd
import requests
from bs4 import BeautifulSoup

# Part 1 - Create DataFrame with PostalCode, Borough, Neighborhood

### Importing Data

In [2]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
html_data = requests.get(url).text

In [3]:
soup = BeautifulSoup(html_data, 'html5lib')

In [4]:
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

In [5]:
# print(table_contents)
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

In [6]:
df.head()

PostalCode           Borough                      Neighborhood
0        M3A        North York                         Parkwoods
1        M4A        North York                  Victoria Village
2        M5A  Downtown Toronto         Regent Park, Harbourfront
3        M6A        North York  Lawrence Manor, Lawrence Heights
4        M7A      Queen's Park     Ontario Provincial Government

Response 200 means that we are able to make the connection to the page

Group df by PostalCode, reset index, and remove "Index" column

In [7]:
df = df.groupby(['PostalCode']).head()
df.head()

PostalCode           Borough                      Neighborhood
0        M3A        North York                         Parkwoods
1        M4A        North York                  Victoria Village
2        M5A  Downtown Toronto         Regent Park, Harbourfront
3        M6A        North York  Lawrence Manor, Lawrence Heights
4        M7A      Queen's Park     Ontario Provincial Government

In [8]:
df = df.reset_index(drop=True)
df

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A           Queen's Park   
..         ...                    ...   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4                        Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [9]:
df.shape

(103, 3)

### Part 1 Answer: 103 rows and 3 columns

# Part 2 - Latitude and Lognitude Coordinates for Neighborhoods

### Required packages for assignment

In [10]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

check the shape of the two dataframes we need to join

In [11]:
df.shape

(103, 3)

In [12]:
data.shape

(103, 3)

Let's make sure the Postal Code columns are the same data type

In [13]:
df.dtypes

PostalCode      object
Borough         object
Neighborhood    object
dtype: object

In [14]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [15]:
corr_data = df.join(data.set_index('Postal Code'), on='PostalCode', how='inner')
corr_data.head()

PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

In [16]:
corr_data.shape

(103, 5)

### Part 2 Answer: 103 rows and 5 columns

# Part 3 - Explore and Cluster Neighborhoods in Toronto

### Required packages for assignment

In [22]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

import matplotlib.cm as cm
import matplotlib.colors as colors

print('Libraries imported.')

Libraries imported.


## A. Explore Data Set

The following analysis is based on the "Segmenting and Clustering Neighborhoods in New York City" assignment from Week 3 in the Applied Data Science Capstone

In [32]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geographical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geographical coordinates of Toronto are 43.6534817, -79.3839347.


#### Map of Toronto with Neighborhoods

In [28]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighborhood in zip(corr_data['Latitude'], corr_data['Longitude'], corr_data['Borough'], corr_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

## B. Neighborhood Venues in Toronto

#### Define Foursquare Credentials and Version

In [30]:
CLIENT_ID = 'VKBN2JB0UTHRNRDT40G3EUN0FKLUUU11NBNGPVJFVYET300I' 
CLIENT_SECRET = '1FMHFZZFILBNV2EKSRACRGKU0RA1Z2WLDIYF4J40KTRK3RDW'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: VKBN2JB0UTHRNRDT40G3EUN0FKLUUU11NBNGPVJFVYET300I
CLIENT_SECRET:1FMHFZZFILBNV2EKSRACRGKU0RA1Z2WLDIYF4J40KTRK3RDW


Function to get venues for all neighborhoods in Toronto

In [39]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Collecting the venues in Toronto for each Neighbourhood

In [40]:
venues_toronto = getNearbyVenues(corr_data['Neighborhood'], corr_data['Latitude'], corr_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills North
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview East
The Danforth

In [41]:
venues_toronto.shape

(1326, 5)

venues are collected for all neighborhoods, total = 1326


In [42]:
venues_toronto.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2  Victoria Village              43.725882              -79.315572   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2  Victoria Village Arena           Hockey Arena  
3               Portugril  Portuguese Restaurant  
4             Tim Hortons            Coffee Shop

Let's check how many venues were returned for each neighborhood

In [44]:
venues_toronto.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               5   
Alderwood, Long Branch                                                  7   
Bathurst Manor, Wilson Heights, Downsview North                        22   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      25   
Berczy Park                                                            30   
Birch Cliff, Cliffside West                                             4   
Brockton, Parkdale Village, Exhibition Place                           23   
CN Tower, King and Spadina, Railway Lands, Harb...                     12   
Caledonia-Fairbanks                                                     4   
Cedarbrae                                                               8   
Central Bay Street                                                     30   
Christie                                                               16   
Church and Wellesley                                                   30   
Clairville, Humberwood, Woodbine Downs, West Hu...                      5   
Clarks Corners, Tam O'Shanter, Sullivan                                12   
Cliffside, Cliffcrest, Scarborough Village West                         3   
Commerce Court, Victoria Hotel                                         30   
Davisville                                                             30   
Davisville North                                                        9   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        4   
Don Mills North                                                         5   
Don Mills South                                                        19   
Dorset Park, Wexford Heights, Scarborough Town ...                      6   
Downsview Central                                                       3   
Downsview East                                                          2   
Downsview Northwest                                                     5   
Downsview West                                                          6   
Dufferin, Dovercourt Village                                           17   
Enclave of L4W                                                         13   
Enclave of M4L                                                         17   
Enclave of M5E                                                         30   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                      7   
Fairview, Henry Farm, Oriole                                           30   
First Canadian Place, Underground city                                 30   
Forest Hill North & West                                                4   
Garden District, Ryerson                                               30   
Glencairn                                                               5   
Golden Mile, Clairlea, Oakridge                                        10   
Guildwood, Morningside, West Hill                                       8   
Harbourfront East, Union Station, Toronto Islands                      30   
High Park, The Junction South                                          26   
Hillcrest Village                                                       5   
Humber Summit                                                           1   
Humberlea, Emery                                                        2   
Humewood-Cedarvale                                                      5   
India Bazaar, The Beaches West                                         21   
Kennedy Park, Ionview, East Birchmount Park                             5   
Kensington Market, Chinatown, Grange Park                              30   
Kingsview Village, St. Phillips, Martin Grove G...                      4   
Lawrence Manor, Lawrence Heights                

#### Let's find out how many unique categories can be curated from all the returned venues

In [46]:
print('There are {} uniques categories.'.format(len(venues_toronto['Venue Category'].unique())))

There are 230 uniques categories.


## C. Analyze Each Neighborhood

#### One Hot encoding the venue Categories

In [112]:
# one hot encoding
toronto_onehot = pd.get_dummies(venues_toronto[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = venues_toronto['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Adult Boutique  Airport  \
0            0                  0               0        0   
1            0                  0               0        0   
2            0                  0               0        0   
3            0                  0               0        0   
4            0                  0               0        0   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0               0                0                 0   
1                   0               0                0                 0   
2                   0               0                0                 0   
3                   0               0                0                 0   
4                   0               0                0                 0   

   American Restaurant  Antique Shop  Art Gallery  Art Museum  \
0                    0             0            0           0   
1                    0             0            0           0   
2                    0             0            0           0   
3                    0             0            0           0   
4                    0             0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0                    0                 0                   0            0   
1                    0                 0                   0            0   
2                    0                 0                   0            0   
3                    0                 0                   0            0   
4                    0                 0                   0            0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Field  Basketball Court  Basketball Stadium  Beer Bar  Beer Store  \
0               0                 0                   0         0           0   
1               0                 0                   0         0           0   
2               0                 0                   0         0           0   
3               0                 0                   0         0           0   
4               0                 0                   0         0           0   

   Belgian Restaurant  Bike Shop  Bistro  Board Shop  Boat or Ferry  \
0                   0          0       0           0              0   
1                   0          0       0           0              0   
2                   0          0       0           0              0   
3                   0          0       0           0              0   
4                   0          0       0           0              0   

   Bookstore  Boutique  Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  \
0          0         0               0        0            0                0   
1          0         0               0        0            0                0   
2          0         0               0        0            0                0   
3          0         0               0        0            0                0   
4          0         0               0        0            0                0   

   Burger Joint  Burrito Place  Bus Line  Bus Station  Business Service  \
0             0              0         0            0                 0   
1             0              0         0            0                 0   
2             0              0         0            0                 0   
3             0              0         0            0                 0   
4             0              0         0            0                 0   

   Butcher  Café  Caju

And let's examine the new dataframe size.

In [94]:
toronto_onehot.shape

(1326, 230)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [95]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                           Agincourt     0.000000   
1                              Alderwood, Long Branch     0.000000   
2     Bathurst Manor, Wilson Heights, Downsview North     0.000000   
3                                     Bayview Village     0.000000   
4                   Bedford Park, Lawrence Manor East     0.000000   
5                                         Berczy Park     0.000000   
6                         Birch Cliff, Cliffside West     0.000000   
7        Brockton, Parkdale Village, Exhibition Place     0.000000   
8   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
9                                 Caledonia-Fairbanks     0.000000   
10                                          Cedarbrae     0.000000   
11                                 Central Bay Street     0.033333   
12                                           Christie     0.000000   
13                               Church and Wellesley     0.000000   
14  Clairville, Humberwood, Woodbine Downs, West H...     0.000000   
15            Clarks Corners, Tam O'Shanter, Sullivan     0.000000   
16    Cliffside, Cliffcrest, Scarborough Village West     0.000000   
17                     Commerce Court, Victoria Hotel     0.000000   
18                                         Davisville     0.000000   
19                                   Davisville North     0.000000   
20   Del Ray, Mount Dennis, Keelsdale and Silverthorn     0.000000   
21                                    Don Mills North     0.000000   
22                                    Don Mills South     0.000000   
23  Dorset Park, Wexford Heights, Scarborough Town...     0.000000   
24                                  Downsview Central     0.000000   
25                                     Downsview East     0.000000   
26                                Downsview Northwest     0.000000   
27                                     Downsview West     0.000000   
28                       Dufferin, Dovercourt Village     0.000000   
29                                     Enclave of L4W     0.000000   
30                                     Enclave of M4L     0.000000   
31                                     Enclave of M5E     0.000000   
32  Eringate, Bloordale Gardens, Old Burnhamthorpe...     0.000000   
33                       Fairview, Henry Farm, Oriole     0.000000   
34             First Canadian Place, Underground city     0.000000   
35                           Forest Hill North & West     0.000000   
36                           Garden District, Ryerson     0.000000   
37                                          Glencairn     0.000000   
38                    Golden Mile, Clairlea, Oakridge     0.000000   
39                  Guildwood, Morningside, West Hill     0.000000   
40  Harbourfront East, Union Station, Toronto Islands     0.000000   
41                      High Park, The Junction South     0.000000   
42                                  Hillcrest Village     0.000000   
43                                      Humber Summit     0.000000   
44                                   Humberlea, Emery     0.000000   
45                                 Humewood-Cedarvale     0.000000   
46                     India Bazaar, The Beaches West     0.000000   
47        Kennedy Park, Ionview, East Birchmount Park     0.000000   
48          Kensington Market, Chinatown, Grange Park     0.000000   
49  Kingsview Village, St. Phillips, Martin Grove ...     0.000000   
50                   Lawrence Manor, Lawrence Heights     0.000000   
51                                      Lawrence Park     0.000000   
52                                            Leaside     0.000000   
53                           Little Portugal, Trinity     0.033333   
54                                     Malvern, Rouge     0.000000   
55  Milliken, Agincourt North, Steeles East, L'Amo...     0.000000   
56  Mimico NW, The Queensway West, South of Bloor,.

#### Let's confirm the new size

In [96]:
toronto_grouped.shape

(100, 230)

#### Let's print each neighborhood along with the top 5 most common venues

In [97]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0                     Lounge   0.2
1               Skating Rink   0.2
2             Clothing Store   0.2
3  Latin American Restaurant   0.2
4             Breakfast Spot   0.2


----Alderwood, Long Branch----
          venue  freq
0   Pizza Place  0.29
1           Pub  0.14
2           Gym  0.14
3   Coffee Shop  0.14
4  Skating Rink  0.14


----Bathurst Manor, Wilson Heights, Downsview North----
               venue  freq
0        Coffee Shop  0.09
1               Bank  0.09
2  Health Food Store  0.05
3      Deli / Bodega  0.05
4      Shopping Mall  0.05


----Bayview Village----
                 venue  freq
0   Chinese Restaurant  0.25
1                 Bank  0.25
2                 Café  0.25
3  Japanese Restaurant  0.25
4          Yoga Studio  0.00


----Bedford Park, Lawrence Manor East----
                  venue  freq
0        Sandwich Place  0.08
1           Coffee Shop  0.08
2    Italian Restaurant  0.08
3  Fast Food Restaurant

               venue  freq
0               Park  0.25
1     Sandwich Place  0.25
2           Bus Line  0.25
3  Mobile Phone Shop  0.25
4        Yoga Studio  0.00


----Lawrence Manor, Lawrence Heights----
                    venue  freq
0          Clothing Store  0.33
1       Accessories Store  0.22
2  Furniture / Home Store  0.11
3                Boutique  0.11
4   Vietnamese Restaurant  0.11


----Lawrence Park----
              venue  freq
0  Business Service  0.25
1              Park  0.25
2       Swim School  0.25
3          Bus Line  0.25
4       Yoga Studio  0.00


----Leaside----
                    venue  freq
0     Sporting Goods Shop  0.10
1             Coffee Shop  0.10
2            Burger Joint  0.07
3                    Bank  0.07
4  Furniture / Home Store  0.07


----Little Portugal, Trinity----
                   venue  freq
0                    Bar  0.13
1  Vietnamese Restaurant  0.07
2            Coffee Shop  0.07
3       Asian Restaurant  0.07
4            Yoga Studi

                   venue  freq
0            Coffee Shop  0.67
1  Korean BBQ Restaurant  0.33
2            Yoga Studio  0.00
3                  Motel  0.00
4         Massage Studio  0.00


----Woodbine Heights----
                venue  freq
0        Skating Rink  0.29
1                Park  0.14
2          Beer Store  0.14
3  Athletics & Sports  0.14
4        Dance Studio  0.14


----York Mills West----
                 venue  freq
0                 Park  0.67
1    Convenience Store  0.33
2          Yoga Studio  0.00
3  Monument / Landmark  0.00
4  Martial Arts School  0.00




#### Let's put that into a _pandas_ dataframe

First, let's write a function to sort the venues in descending order.

In [174]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [175]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

In [176]:
for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

In [177]:
neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village    Chinese Restaurant   
4                Bedford Park, Lawrence Manor East        Sandwich Place   

  2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
0          Skating Rink        Clothing Store  Latin American Restaurant   
1                   Pub                   Gym                Coffee Shop   
2                  Bank     Health Food Store              Deli / Bodega   
3                  Bank                  Café        Japanese Restaurant   
4           Coffee Shop    Italian Restaurant       Fast Food Restaurant   

  5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
0        Breakfast Spot           Yoga Studio        Monument / Landmark   
1          Skating Rink        Sandwich Place                Yoga Studio   
2         Shopping Mall        Sandwich Place  Middle Eastern Restaurant   
3           Yoga Studio                 Motel             Medical Center   
4               Butcher                  Café               Liquor Store   

      8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0  Mediterranean Restaurant                Men's Store          Metro Station  
1        Miscellaneous Shop  Middle Eastern Restaurant     Mexican Restaurant  
2         Mobile Phone Shop                 Restaurant            Supermarket  
3  Mediterranean Restaurant                Men's Store          Metro Station  
4                Restaurant                  Juice Bar       Sushi Restaurant

## D. Cluster Neighborhoods

Run _k_-means to cluster the neighborhood into 5 clusters.

In [178]:
# set number of clusters
kclusters = 10

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([8, 1, 1, 8, 1, 1, 8, 1, 8, 5], dtype=int32)

In [179]:
toronto_grouped_clustering.head()

Yoga Studio  Accessories Store  Adult Boutique  Airport  \
0          0.0                0.0             0.0      0.0   
1          0.0                0.0             0.0      0.0   
2          0.0                0.0             0.0      0.0   
3          0.0                0.0             0.0      0.0   
4          0.0                0.0             0.0      0.0   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                 0.0             0.0              0.0               0.0   
1                 0.0             0.0              0.0               0.0   
2                 0.0             0.0              0.0               0.0   
3                 0.0             0.0              0.0               0.0   
4                 0.0             0.0              0.0               0.0   

   American Restaurant  Antique Shop  Art Gallery  Art Museum  \
0                 0.00           0.0          0.0         0.0   
1                 0.00           0.0          0.0         0.0   
2                 0.00           0.0          0.0         0.0   
3                 0.00           0.0          0.0         0.0   
4                 0.04           0.0          0.0         0.0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0                  0.0               0.0                 0.0          0.0   
1                  0.0               0.0                 0.0          0.0   
2                  0.0               0.0                 0.0          0.0   
3                  0.0               0.0                 0.0          0.0   
4                  0.0               0.0                 0.0          0.0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery      Bank  Bar  \
0            0.0        0.0         0.0         0.0     0.0  0.000000  0.0   
1            0.0        0.0         0.0         0.0     0.0  0.000000  0.0   
2            0.0        0.0         0.0         0.0     0.0  0.090909  0.0   
3            0.0        0.0         0.0         0.0     0.0  0.250000  0.0   
4            0.0        0.0         0.0         0.0     0.0  0.000000  0.0   

   Baseball Field  Basketball Court  Basketball Stadium  Beer Bar  Beer Store  \
0             0.0               0.0                 0.0       0.0         0.0   
1             0.0               0.0                 0.0       0.0         0.0   
2             0.0               0.0                 0.0       0.0         0.0   
3             0.0               0.0                 0.0       0.0         0.0   
4             0.0               0.0                 0.0       0.0         0.0   

   Belgian Restaurant  Bike Shop  Bistro  Board Shop  Boat or Ferry  \
0                 0.0        0.0     0.0         0.0            0.0   
1                 0.0        0.0     0.0         0.0            0.0   
2                 0.0        0.0     0.0         0.0            0.0   
3                 0.0        0.0     0.0         0.0            0.0   
4                 0.0        0.0     0.0         0.0            0.0   

   Bookstore  Boutique  Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  \
0        0.0       0.0             0.2      0.0     0.000000              0.0   
1        0.0       0.0             0.0      0.0     0.000000              0.0   
2        0.0       0.0             0.0      0.0     0.045455              0.0   
3        0.0       0.0             0.0      0.0     0.000000              0.0   
4        0.0       0.0             0.0      0.0     0.000000              0.0   

   Burger Joint  Burrito Place  Bus Line  Bus Station  Business Service  \
0           0.0            0.0       0.0          0.0               0.0   
1           0.0            0.0       0.0          0.0               0.0   
2           0.0            0.0       0.0          0.0               0.0   
3           0.0            0.0       0.0          0.0               0.0   
4           0.0            0.0       0.0          0.0               0.0   

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [180]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [181]:
neighborhoods_venues_sorted.head()

Cluster Labels                                     Neighborhood  \
0               8                                        Agincourt   
1               1                           Alderwood, Long Branch   
2               1  Bathurst Manor, Wilson Heights, Downsview North   
3               8                                  Bayview Village   
4               1                Bedford Park, Lawrence Manor East   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                Lounge          Skating Rink        Clothing Store   
1           Pizza Place                   Pub                   Gym   
2           Coffee Shop                  Bank     Health Food Store   
3    Chinese Restaurant                  Bank                  Café   
4        Sandwich Place           Coffee Shop    Italian Restaurant   

       4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0  Latin American Restaurant        Breakfast Spot           Yoga Studio   
1                Coffee Shop          Skating Rink        Sandwich Place   
2              Deli / Bodega         Shopping Mall        Sandwich Place   
3        Japanese Restaurant           Yoga Studio                 Motel   
4       Fast Food Restaurant               Butcher                  Café   

       7th Most Common Venue     8th Most Common Venue  \
0        Monument / Landmark  Mediterranean Restaurant   
1                Yoga Studio        Miscellaneous Shop   
2  Middle Eastern Restaurant         Mobile Phone Shop   
3             Medical Center  Mediterranean Restaurant   
4               Liquor Store                Restaurant   

       9th Most Common Venue 10th Most Common Venue  
0                Men's Store          Metro Station  
1  Middle Eastern Restaurant     Mexican Restaurant  
2                 Restaurant            Supermarket  
3                Men's Store          Metro Station  
4                  Juice Bar       Sushi Restaurant

In [182]:
toronto_merged = corr_data.drop(['PostalCode'], axis=1)

In [183]:
# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

In [184]:
toronto_merged.head() # check the last columns!

Borough                      Neighborhood   Latitude  Longitude  \
0        North York                         Parkwoods  43.753259 -79.329656   
1        North York                  Victoria Village  43.725882 -79.315572   
2  Downtown Toronto         Regent Park, Harbourfront  43.654260 -79.360636   
3        North York  Lawrence Manor, Lawrence Heights  43.718518 -79.464763   
4      Queen's Park     Ontario Provincial Government  43.662301 -79.389494   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0             0.0                  Park     Food & Drink Shop   
1             1.0           Pizza Place     French Restaurant   
2             1.0           Coffee Shop                  Park   
3             1.0        Clothing Store     Accessories Store   
4             1.0           Coffee Shop      Sushi Restaurant   

    3rd Most Common Venue       4th Most Common Venue  5th Most Common Venue  \
0             Yoga Studio  Modern European Restaurant                 Market   
1             Coffee Shop                Hockey Arena  Portuguese Restaurant   
2                  Bakery              Breakfast Spot                   Café   
3  Furniture / Home Store                    Boutique  Vietnamese Restaurant   
4             Yoga Studio          Mexican Restaurant          Smoothie Shop   

  6th Most Common Venue      7th Most Common Venue 8th Most Common Venue  \
0   Martial Arts School             Massage Studio        Medical Center   
1        Medical Center   Mediterranean Restaurant           Men's Store   
2            Restaurant             Chocolate Shop                   Pub   
3           Coffee Shop  Middle Eastern Restaurant   Monument / Landmark   
4        Sandwich Place                       Café              Beer Bar   

        9th Most Common Venue 10th Most Common Venue  
0    Mediterranean Restaurant            Men's Store  
1               Metro Station     Mexican Restaurant  
2       Performing Arts Venue            Yoga Studio  
3  Modern European Restaurant      Mobile Phone Shop  
4         Fried Chicken Joint             Restaurant

Remove NAN values for neighborhoods without venues

In [185]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Finally, let's visualize the resulting clusters

In [186]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighborhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Finally, let's visualize the resulting clusters

In [197]:
toronto_merged_nonan['Cluster Labels']

0      0.0
1      1.0
2      1.0
3      1.0
4      1.0
6      9.0
7      8.0
8      8.0
9      1.0
10     8.0
11     2.0
12     8.0
13     1.0
14     8.0
15     1.0
16     8.0
17     1.0
18     8.0
19     8.0
20     1.0
21     5.0
22     1.0
23     1.0
24     1.0
25     8.0
26     8.0
27     9.0
28     1.0
29     1.0
30     1.0
31     8.0
32     7.0
33     1.0
34     1.0
35     0.0
36     8.0
37     1.0
38     1.0
39     8.0
40     0.0
41     1.0
42     1.0
43     1.0
44     8.0
46     8.0
47     1.0
48     1.0
49     8.0
50     6.0
51     8.0
52     0.0
53     8.0
54     1.0
55     1.0
56     9.0
57     4.0
58     8.0
59     1.0
60     8.0
61     8.0
62     3.0
63     8.0
64     0.0
65     1.0
66     0.0
67     8.0
68     8.0
69     8.0
70     1.0
71     8.0
72     1.0
73     1.0
74     1.0
75     1.0
76     1.0
77     8.0
78     8.0
79     1.0
80     8.0
81     1.0
82     1.0
83     7.0
84     8.0
85     7.0
86     1.0
87     8.0
88     8.0
89     1.0
90     1.0
91     7.0
92     8.0

## E. Examine Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.

#### Cluster 1

In [198]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0                 Parkwoods                  Park     Food & Drink Shop   
35       The Danforth  East                  Park     Convenience Store   
40           Downsview East               Airport                  Park   
52  Willowdale, Newtonbrook                  Park           Yoga Studio   
64                   Weston                  Park           Yoga Studio   
66          York Mills West                  Park     Convenience Store   

   3rd Most Common Venue       4th Most Common Venue 5th Most Common Venue  \
0            Yoga Studio  Modern European Restaurant                Market   
35         Metro Station                 Yoga Studio   Monument / Landmark   
40           Yoga Studio         Monument / Landmark   Martial Arts School   
52   Monument / Landmark         Martial Arts School        Massage Studio   
64   Monument / Landmark         Martial Arts School        Massage Studio   
66           Yoga Studio         Monument / Landmark   Martial Arts School   

   6th Most Common Venue     7th Most Common Venue     8th Most Common Venue  \
0    Martial Arts School            Massage Studio            Medical Center   
35   Martial Arts School            Massage Studio            Medical Center   
40        Massage Studio            Medical Center  Mediterranean Restaurant   
52        Medical Center  Mediterranean Restaurant               Men's Store   
64        Medical Center  Mediterranean Restaurant               Men's Store   
66        Massage Studio            Medical Center  Mediterranean Restaurant   

       9th Most Common Venue 10th Most Common Venue  
0   Mediterranean Restaurant            Men's Store  
35  Mediterranean Restaurant            Men's Store  
40               Men's Store          Metro Station  
52             Metro Station     Mexican Restaurant  
64             Metro Station     Mexican Restaurant  
66               Men's Store          Metro Station

#### Cluster 2

In [199]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Neighborhood 1st Most Common Venue  \
1                                    Victoria Village           Pizza Place   
2                           Regent Park, Harbourfront           Coffee Shop   
3                    Lawrence Manor, Lawrence Heights        Clothing Store   
4                       Ontario Provincial Government           Coffee Shop   
9                            Garden District, Ryerson                  Café   
13                                    Don Mills South           Coffee Shop   
15                                     St. James Town                  Café   
17  Eringate, Bloordale Gardens, Old Burnhamthorpe...              Pharmacy   
20                                        Berczy Park        Farmers Market   
22                                             Woburn           Coffee Shop   
23                                            Leaside   Sporting Goods Shop   
24                                 Central Bay Street           Coffee Shop   
28    Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
29                                   Thorncliffe Park     Indian Restaurant   
30                           Richmond, Adelaide, King           Coffee Shop   
33                       Fairview, Henry Farm, Oriole        Clothing Store   
34                    Northwood Park, York University        Massage Studio   
37                           Little Portugal, Trinity                   Bar   
38        Kennedy Park, Ionview, East Birchmount Park        Discount Store   
41                       The Danforth West, Riverdale      Greek Restaurant   
42           Toronto Dominion Centre, Design Exchange                  Café   
43       Brockton, Parkdale Village, Exhibition Place                  Café   
47                     India Bazaar, The Beaches West  Fast Food Restaurant   
48                     Commerce Court, Victoria Hotel                  Café   
54                                    Studio District           Coffee Shop   
55                  Bedford Park, Lawrence Manor East        Sandwich Place   
59                                   Willowdale South      Ramen Restaurant   
65  Dorset Park, Wexford Heights, Scarborough Town...     Indian Restaurant   
70                                          Westmount           Pizza Place   
72                                    Willowdale West         Grocery Store   
73                                 North Toronto West           Coffee Shop   
74                The Annex, North Midtown, Yorkville        Sandwich Place   
75                             Parkdale, Roncesvalles             Gift Shop   
76                                     Enclave of L4W           Coffee Shop   
79                                         Davisville          Dessert Shop   
81                                 Runnymede, Swansea           Pizza Place   
82            Clarks Corners, Tam O'Shanter, Sullivan           Pizza Place   
86  Summerhill West, Rathnelly, South Hill, Forest...           Coffee Shop   
89  South Steeles, Silverstone, Humbergate, Jamest...         Grocery Store   
90                      Steeles West, L'Amoreaux West  Fast Food Restaurant   
93                             Alderwood, Long Branch           Pizza Place   
96                        St. James Town, Cabbagetown                Bakery   
97             First Canadian Place, Underground city                  Café   
99                               Church and Wellesley        Breakfast Spot   

    2nd Most Common Venue        3rd Most Common Venue 4th Most Common Venue  \
1       French Restaurant                  Coffee Shop          Hockey Arena   
2                    Park                       Bakery        Breakfast Spot   
3       Accessories Store       Furniture / Home Store              Boutique   
4        Sushi Restaurant                  Yoga Studio    Mexican Restaurant   
9                 Theater               Clothing Store                Bakery   
13   

#### Cluster 3

In [200]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Neighborhood 1st Most Common Venue  \
11  West Deane Park, Princess Gardens, Martin Grov...                Bakery   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
11           Yoga Studio   Monument / Landmark   Martial Arts School   

   5th Most Common Venue 6th Most Common Venue     7th Most Common Venue  \
11        Massage Studio        Medical Center  Mediterranean Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
11           Men's Store         Metro Station     Mexican Restaurant

#### Cluster 4

In [201]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
62     Roselawn          Home Service                Garden   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
62           Yoga Studio   Monument / Landmark   Martial Arts School   

   6th Most Common Venue 7th Most Common Venue     8th Most Common Venue  \
62        Massage Studio        Medical Center  Mediterranean Restaurant   

   9th Most Common Venue 10th Most Common Venue  
62           Men's Store          Metro Station

#### Cluster 5

In [202]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Neighborhood          1st Most Common Venue 2nd Most Common Venue  \
57  Humberlea, Emery  Paper / Office Supplies Store        Baseball Field   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
57           Yoga Studio   Monument / Landmark   Martial Arts School   

   6th Most Common Venue 7th Most Common Venue     8th Most Common Venue  \
57        Massage Studio        Medical Center  Mediterranean Restaurant   

   9th Most Common Venue 10th Most Common Venue  
57           Men's Store          Metro Station

#### Cluster 6

In [211]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Neighborhood          1st Most Common Venue 2nd Most Common Venue  \
57  Humberlea, Emery  Paper / Office Supplies Store        Baseball Field   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
57           Yoga Studio   Monument / Landmark   Martial Arts School   

   6th Most Common Venue 7th Most Common Venue     8th Most Common Venue  \
57        Massage Studio        Medical Center  Mediterranean Restaurant   

   9th Most Common Venue 10th Most Common Venue  
57           Men's Store          Metro Station

#### Cluster 7

In [212]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 6, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Neighborhood   1st Most Common Venue 2nd Most Common Venue  \
50  Humber Summit  Furniture / Home Store   Monument / Landmark   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
50                Market   Martial Arts School        Massage Studio   

   6th Most Common Venue     7th Most Common Venue 8th Most Common Venue  \
50        Medical Center  Mediterranean Restaurant           Men's Store   

   9th Most Common Venue 10th Most Common Venue  
50         Metro Station     Mexican Restaurant

#### Cluster 8

In [215]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 7, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Neighborhood 1st Most Common Venue  \
32                                Scarborough Village            Playground   
83                        Moore Park, Summerhill East            Restaurant   
85  Milliken, Agincourt North, Steeles East, L'Amo...          Intersection   
91                                           Rosedale                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
32        Cosmetics Shop           Yoga Studio   Monument / Landmark   
83            Playground                  Park   Monument / Landmark   
85            Playground                  Park           Yoga Studio   
91            Playground                 Trail           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
32   Martial Arts School        Massage Studio        Medical Center   
83   Martial Arts School        Massage Studio        Medical Center   
85   Martial Arts School        Massage Studio        Medical Center   
91   Monument / Landmark   Martial Arts School        Massage Studio   

       8th Most Common Venue     9th Most Common Venue 10th Most Common Venue  
32  Mediterranean Restaurant               Men's Store          Metro Station  
83  Mediterranean Restaurant               Men's Store          Metro Station  
85  Mediterranean Restaurant               Men's Store          Metro Station  
91            Medical Center  Mediterranean Restaurant            Men's Store

#### Cluster 9

In [217]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 8, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Neighborhood 1st Most Common Venue  \
7                                      Don Mills North                   Gym   
8                      Parkview Hill, Woodbine Gardens           Pizza Place   
10                                           Glencairn           Pizza Place   
12              Rouge Hill, Port Union, Highland Creek                   Bar   
14                                    Woodbine Heights          Skating Rink   
16                                  Humewood-Cedarvale          Tennis Court   
18                   Guildwood, Morningside, West Hill          Intersection   
19                                         The Beaches     Health Food Store   
25                                            Christie         Grocery Store   
26                                           Cedarbrae      Hakka Restaurant   
31                        Dufferin, Dovercourt Village                Bakery   
36   Harbourfront East, Union Station, Toronto Islands                 Plaza   
39                                     Bayview Village    Chinese Restaurant   
44                     Golden Mile, Clairlea, Oakridge              Bus Line   
46                                      Downsview West         Grocery Store   
49            North Park, Maple Leaf Park, Upwood Park                  Park   
51     Cliffside, Cliffcrest, Scarborough Village West                 Motel   
53                                   Downsview Central        Baseball Field   
58                         Birch Cliff, Cliffside West       College Stadium   
60                                 Downsview Northwest         Grocery Store   
61                                       Lawrence Park      Business Service   
63                       Runnymede, The Junction North         Grocery Store   
67                                    Davisville North           Pizza Place   
68                            Forest Hill North & West                  Park   
69                       High Park, The Junction South    Mexican Restaurant   
71                                   Wexford, Maryvale           Auto Garage   
77   Kingsview Village, St. Phillips, Martin Grove ...                  Park   
78                                           Agincourt                Lounge   
80                      University of Toronto, Harbord                  Café   
84           Kensington Market, Chinatown, Grange Park                  Café   
87   CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
88        New Toronto, Mimico South, Humber Bay Shores           Pizza Place   
92                                      Enclave of M5E        Farmers Market   
94   Clairville, Humberwood, Woodbine Downs, West H...   Rental Car Location   
100                                     Enclave of M4L    Light Rail Station   
102  Mimico NW, The Queensway West, South of Bloor,...         Grocery Store   

          2nd Most Common Venue 3rd Most Common Venue  \
7          Caribbean Restaurant   Japanese Restaurant   
8          Gym / Fitness Center             Gastropub   
10                         Park                Bakery   
12                Moving Target           Yoga Studio   
14                         Park            Beer Store   
16                        Trail                 Field   
18            Electronics Store        Medical Center   
19                        Trail                   Pub   
25                         Café                  Park   
26          Fried Chicken Joint                  Bank   
31                     Pharmacy         Grocery Store   
36                         Café                  Park   
39                         Bank                  Café   
44                       Bakery          Intersection   
46                         Park         Shopping Mall   
49   Construction & Landscaping      Basketball Court   
51                 Skating Rink   American Restaurant   
53                   Food Truck      Business Service  

#### Cluster 10

In [220]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 9, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Neighborhood 1st Most Common Venue  \
6                                     Malvern, Rouge  Fast Food Restaurant   
27                                 Hillcrest Village  Fast Food Restaurant   
56  Del Ray, Mount Dennis, Keelsdale and Silverthorn            Restaurant   

   2nd Most Common Venue       3rd Most Common Venue  \
6            Yoga Studio  Modern European Restaurant   
27           Golf Course                        Pool   
56  Fast Food Restaurant              Discount Store   

       4th Most Common Venue       5th Most Common Venue  \
6                     Market         Martial Arts School   
27  Mediterranean Restaurant                     Dog Run   
56            Sandwich Place  Modern European Restaurant   

   6th Most Common Venue       7th Most Common Venue  \
6         Massage Studio              Medical Center   
27           Yoga Studio  Modern European Restaurant   
56        Massage Studio              Medical Center   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
6   Mediterranean Restaurant           Men's Store          Metro Station  
27       Martial Arts School        Massage Studio         Medical Center  
56  Mediterranean Restaurant           Men's Store          Metro Station

Toronto neighborhoods did not cluster well with only 5.  The majority of the neighborhoods clustered together with a few outliers.  The number was increased to 7 then 10.  The final cluster of 10 provides enough separation of the neighborhoods to be useful.

### This project was completed!